test/extension/server/server.go

```diff
@@ -0,0 +1,292 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package server contains the implementation of a RuntimeSDK webhook server.
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	runtimecatalog "sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+// DefaultPort is the default port that the webhook server serves.
+var DefaultPort = 9443
+
+// Server is a runtime webhook server.
+type Server struct {
+	catalog  *runtimecatalog.Catalog
+	server   *webhook.Server
+	handlers map[string]ExtensionHandler
+}
+
+// Options are the options for the Server.
+type Options struct {
+	// Catalog is the catalog used to handle requests.
+	Catalog *runtimecatalog.Catalog
+
+	// Port is the port that the webhook server serves at.
+	// It is used to set webhook.Server.Port.
+	Port int
+
+	// Host is the hostname that the webhook server binds to.
+	// It is used to set webhook.Server.Host.
+	Host string
+
+	// CertDir is the directory that contains the server key and certificate.
+	// If not set, webhook server would look up the server key and certificate in
+	// {TempDir}/k8s-webhook-server/serving-certs. The server key and certificate
+	// must be named tls.key and tls.crt, respectively.
+	// It is used to set webhook.Server.CertDir.
+	CertDir string
+}
+
+// NewServer creates a new runtime webhook server based on the given Options.
+func NewServer(options Options) (*Server, error) {
+	if options.Catalog == nil {
+		return nil, errors.Errorf("catalog is required")
+	}
+	if options.Port <= 0 {
+		options.Port = DefaultPort
+	}
+	if options.CertDir == "" {
+		options.CertDir = filepath.Join(os.TempDir(), "k8s-webhook-server", "serving-certs")
+	}
+
+	webhookServer := &webhook.Server{
+		Port:          options.Port,
+		Host:          options.Host,
+		CertDir:       options.CertDir,
+		CertName:      "tls.crt",
+		KeyName:       "tls.key",
+		WebhookMux:    http.NewServeMux(),
+		TLSMinVersion: "1.3",
+	}
+
+	return &Server{
+		catalog:  options.Catalog,
+		server:   webhookServer,
+		handlers: map[string]ExtensionHandler{},
+	}, nil
+}
+
+// ExtensionHandler represents an extension handler.
+type ExtensionHandler struct {
```

Comment:
Q: Should we rename this? I was confused because we also have `runtimev1.ExtensionHandler` and `runtimehooksv1.ExtensionHandler` and now here also an `ExtensionHandler` struct.

I consider it ok because it is a test package in the test go module and not in the core capi module, also it is not meant to get imported/used elsewhere **currently** except tests in here.

Reply:
Maybe something like `HTTPExtensionHandler` because it includes the HandlerFunc`?

Reply:
Hm. I would prefer avoiding to hard-code this to http. As this struct today is entirely independent of HTTP and could be also plugged into something else than an HTTP server.

Reply:
I can't think of a better name which represents what this struct is. 
Except if we have a good idea, I would personally prefer keeping the name. I prefer having a fitting name over deduplicating struct names across packages.

A RuntimeExtension author only has to ever work with this struct. They are not excepted to write DiscoveryResponses or read the status of the ExtensionConfig CRD.

Reply:
As mentioned: I'm also okay with keeping :-)